mediagrains_py36/numpy/convert.py

```diff
@@ -0,0 +1,275 @@
+
+# Bit depth conversions
+def _unbiased_right_shift(a: np.ndarray, n: int) -> np.ndarray:
+    return (a >> n) + ((a >> (n - 1))&0x1)
+
+def _bitdepth_down_convert_yuv(grain_in: VIDEOGRAIN, grain_out: VIDEOGRAIN):
+    bitshift = COG_FRAME_FORMAT_ACTIVE_BITS(grain_in.format) - COG_FRAME_FORMAT_ACTIVE_BITS(grain_out.format)
+
+    grain_out.component_data[0][:] = _unbiased_right_shift(grain_in.component_data[0][:], bitshift)
+    grain_out.component_data[1][:] = _unbiased_right_shift(grain_in.component_data[1][:], bitshift)
+    grain_out.component_data[2][:] = _unbiased_right_shift(grain_in.component_data[2][:], bitshift)
+
+def _bitdepth_down_convert_rgb(grain_in: VIDEOGRAIN, grain_out: VIDEOGRAIN):
+    bitshift = COG_FRAME_FORMAT_ACTIVE_BITS(grain_in.format) - COG_FRAME_FORMAT_ACTIVE_BITS(grain_out.format)
+
+    grain_out.component_data.R[:] = _unbiased_right_shift(grain_in.component_data.R[:], bitshift)
+    grain_out.component_data.G[:] = _unbiased_right_shift(grain_in.component_data.G[:], bitshift)
+    grain_out.component_data.B[:] = _unbiased_right_shift(grain_in.component_data.B[:], bitshift)
+
+
+def _noisy_left_shift(a: np.ndarray, n: int) -> np.ndarray:
+    rando = ((npr.random_sample(a.shape) * (1 << n)).astype(a.dtype)) & ((1 << n) - 1)
+    return (a << n) + rando
+
+def _bitdepth_up_convert_yuv(grain_in: VIDEOGRAIN, grain_out: VIDEOGRAIN):
+    bitshift = COG_FRAME_FORMAT_ACTIVE_BITS(grain_out.format) - COG_FRAME_FORMAT_ACTIVE_BITS(grain_in.format)
+
+    dt = grain_out.component_data[0].dtype
+
+    grain_out.component_data[0][:] = _noisy_left_shift(grain_in.component_data[0][:].astype(dt), bitshift)
+    grain_out.component_data[1][:] = _noisy_left_shift(grain_in.component_data[1][:].astype(dt), bitshift)
+    grain_out.component_data[2][:] = _noisy_left_shift(grain_in.component_data[2][:].astype(dt), bitshift)
+
+def _bitdepth_up_convert_rgb(grain_in: VIDEOGRAIN, grain_out: VIDEOGRAIN):
+    bitshift = COG_FRAME_FORMAT_ACTIVE_BITS(grain_out.format) - COG_FRAME_FORMAT_ACTIVE_BITS(grain_in.format)
+
+    dt = grain_out.component_data[0].dtype
+
+    grain_out.component_data.R[:] = _noisy_left_shift(grain_in.component_data.R[:].astype(dt), bitshift)
+    grain_out.component_data.G[:] = _noisy_left_shift(grain_in.component_data.G[:].astype(dt), bitshift)
+    grain_out.component_data.B[:] = _noisy_left_shift(grain_in.component_data.B[:].astype(dt), bitshift)
+
+
+# Colourspace conversions (based on rec.709)
+def _convert_rgb_to_yuv444(grain_in: VIDEOGRAIN, grain_out: VIDEOGRAIN):
+    bd = COG_FRAME_FORMAT_ACTIVE_BITS(grain_out.format)
+    (R, G, B) = (grain_in.component_data.R,
+                    grain_in.component_data.G,
+                    grain_in.component_data.B)
+
+    grain_out.component_data.Y[:,:] = (R*0.2126 +  G*0.7152 + B*0.0722)
+    grain_out.component_data.U[:,:] = (R*-0.114572 - G*0.385428 + B*0.5 + (1 << (bd - 1)))
+    grain_out.component_data.V[:,:] = (R*0.5 - G*0.454153 - B*0.045847  + (1 << (bd - 1)))
```

Comment:
I'm curious where these coefficients came from as I can only find the Y coeff and the coeff in _convert_yuv444_to_rgb. I'm assuming they are the mirror of those from _convert_yuv444_to_rgb. The tests are comparing with the test pattern with max diff 3 and so I guess the coeffs match.

Reply:
I pulled them from Poynton's "Digital Video and HDTV: Standards and Interfaces" (aka. the Orange Book).

Reply:
I'm relatively sure they're just the result of inverting the matrix that goes in the other direction.

Reply:
I happen to have a copy in my desk drawer (not that I can remember anything of it). The equations are on page 315 and 316. It does indeed say they are the inverse of each other. I'm never quite sure what the values should be as the internet says so many different things. Also, whether in this case it matters that these relate to analogue rec709 and not digital for example.

Reply:
I compared with Jim Easterbrook's conversion code and the coefficients are essentially the same and so that provides some extra confidence. The additional step that might want to consider adding is clamping the output values to the allowed range (with the assumption of a range).